### PR TITLE
List authored posts on team member page

### PR DESCRIPTION
**Pull request by MDZHX**
Added a list of authored blog posts on each team member's page.
A working prototype :)

### CODE REVIEW

**Comment by malsf21** [comment]
Looks good so far! I think part of your PR relies on #168 - let me see who's on that right now!

**Comment by MDZHX** [comment]
Thanks for the quick reply! I'll also add some preview card prototypes before #156 is implemented.

**Comment by malsf21** [comment]
Changes are looking good Zihan! I think we may wait on #168 depending on when it's going to happen - do you mind touching base with Shounak and seeing what the ETA on that is looking like?

Also, cc: @krashanoff - any thoughts on time complexity / performance implications? This is not my area of expertise (at least, with Ruby haha)

**Comment by MDZHX** [comment]
Sure! Will DM Shounak on slack.

**Comment by MDZHX** [comment]
I just got a reply from Shounak. He plans to get the basic layout done by this weekend and further polish it afterward. He says that he doesn't have an exact ETA on it right now, though.

**Review by krashanoff — COMMENTED** [review]
> any thoughts on time complexity / performance implications? This is not my area of expertise (at least, with Ruby haha)

Definitely isn't my area of expertise either, but at a glance, I don't think this will kill build time on each dev's environment, but it does see a difference in our GitHub builds (the build for this branch was about 30s slower).

image

**Comment by malsf21** [comment]
> I just got a reply from Shounak. He plans to get the basic layout done by this weekend and further polish it afterward. He says that he doesn't have an exact ETA on it right now, though.

Gotcha, in the interest of time for now then, let's merge this first, then merge it into Shounak's branch and let him handle using the component. I'll leave a formal review in a moment.

> Definitely isn't my area of expertise either, but at a glance, I don't think this will kill build time on each dev's environment, but it does see a difference in our GitHub builds (the build for this branch was about 30s slower).

Where are you seeing the `30s` difference @krashanoff? I'm peeping the actions build and it's still ~ 1s.

File: _layouts/team-member.html [hunk 1]
```diff
@@ -42,12 +42,46 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
...
             {% endif %}
         </div>
     </div>
+
+    <!-- blog posts -->
+    <div class="row">
+        <div class="col">
+            <h2 class="title"> Blog posts that {{ page.firstname }} has written: </h3>
+            <div class="team-links">
+                <ul>
```
Comment (malsf21) [inline _layouts/team-member.html:57]: If you're using `<ul>`, the children should be `<li>` - you can always restyle them to fit your original vibe, but make sure you match things up semantically!

File: _layouts/team-member.html [hunk 2]
```diff
@@ -42,12 +42,46 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
...
+                    {% for post in site.posts %}
+                        {% for author in post.authors %}
+                            {% if author.fname == page.firstname and author.lname == page.lastname %}
+                            <div class="post-preview">
+                                <a href={{post.url}}>
+                                    <img  class="post-preview-img" src="{{site.baseurl}}/{{post.image}}" alt="{{post.alt}}" />
+                                </a>
+                                <div class="post-info">
+                                    <h2 class="title">
```
Comment (malsf21) [inline _layouts/team-member.html:66]: We already have an `<h2>` above; you might want to make this a `<h3>` instead!

File: _layouts/team-member.html [hunk 3]
```diff
@@ -42,12 +42,46 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
...
+                    {% endfor %}    
                 </ul>
             </div>
             {% endif %}
         </div>
     </div>
+
+    <!-- blog posts -->
+    <div class="row">
+        <div class="col">
```
Comment (malsf21) [inline _layouts/team-member.html:54]: Is this supposed to be `.row > .col`? There is no row or column haha

File: _layouts/team-member.html [hunk 4]
```diff
@@ -42,12 +42,46 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
...
+                                    <a class="tla-link" href="{{post.url}}">{{post.title}}</a>
+                                    </h2>
+                                    <p class="title">
+                                    by {{post.author}} <span class="text-normal">on {{ post.date | date: "%b %-d, %Y" }}</span> | category: <a class="tla-link" href="{{site.baseurl}}/blog/category/{{post.category}}/">{{post.category}}</a>
+                                    </p>
+                                    <hr class="divider"/>
+                                    <p class="generated-content">{{post.excerpt}}</p>
+                                </div>
+                                </div>
+                            </section>
```
Comment (malsf21) [inline _layouts/team-member.html:76]: Where is the matching `<section>` tag?

**Review by malsf21 — CHANGES_REQUESTED** [review]
Okay, left some quick quick comments for you @MDZHX! Nothing too big, just some small semantic things 😊

**Comment by krashanoff** [comment]
> Where are you seeing the 30s difference @krashanoff?

In the [Test Jekyll Build](https://github.com/uclaacm/teach-la-website/runs/1747136805) action between the latest commit on this branch and the latest on master. But to reiterate, I saw no detriment on my environment. Further, GitHub actions are a lot slower than Netlify deploys. I'd say this merging this is harmless.

**Comment by malsf21** [comment]
@MDZHX let me know when you have a chance to address the comments, and then I can give you an approving review and we can merge!

File: _layouts/team-member.html (hunk 3 above)
Comment (MDZHX) [reply _layouts/team-member.html:54]: Ah, I thought every section of a page was put into a grid and thus has a row/col. Now I know better of the structure :) thanks!

**Comment by MDZHX** [comment]
@malsf21 Sorry for the late fix! I've been solving a connection problem that prevented me to pull from `master` :( I think it's now ready to merge!

File: _layouts/team-member.html [hunk 5]
```diff
@@ -42,12 +42,47 @@ <h1 class="title">{{ page.firstname }} {{ page.lastname }}
...
-                    {% endfor %}
+                    {% endfor %}    
                 </ul>
             </div>
             {% endif %}
         </div>
     </div>
+
+    <!-- blog posts -->
+    <h3 class="title"> Blog posts that {{ page.firstname }} has written: </h3>
```
Comment (malsf21) [inline _layouts/team-member.html:53]: Hm, so I think this is the inverse of the change that we wanted. Keep in mind that headings and subheadings nest; so if you want a heading to be "within" the blog posts that a user has written, the number should increment by one.

In other words - this should be an `<h2>` tag; the heading inside `<div class="post-info">` should be a `h3`!

**Review by malsf21 — CHANGES_REQUESTED** [review]

**Comment by MDZHX** [comment]
Thanks for pointing that out! I think now it's fixed.

**Review by malsf21 — CHANGES_REQUESTED** [review]
Thanks for the quick change!

One last thing I'll mention - can we add a message and/or hide the entire section if the user hasn't written any blog posts? For example, see

https://deploy-preview-213--acm-teach-la-website.netlify.app/team/aghodsian

After this (I've said this a few times, haha) we'll be good to merge!

**Comment by MDZHX** [comment]
I looked up some docs and implemented this feature by assigning a boolean flag - not sure if that's a clean way to do it. I also removed the outer div with class="team-links" since I don't see this class in the CSS files. Since I'm not familiar enough with the class structure, I wonder if you think I should remove that?

**Comment by malsf21** [comment]
> I looked up some docs and implemented this feature by assigning a boolean flag - not sure if that's a clean way to do it.

Great question! I think your approach works, but after thinking about it for a hot minute I think there's probably a more elegant solution that involves the [where filter](https://jekyllrb.com/docs/liquid/filters/):

```
{% assign authors_posts = site.posts | where_exp:"post", "SOME_EXPRESSION_YOU_NEED_TO_WRITE" | sort_by:"date" %}
```

This will create an iterable `authors_posts` that only has the authors posts. Then, you can check the size; if it's `> 0`, then render the title + iterate through `authors_posts`. This may also make the process faster!

**Comment by MDZHX** [comment]
I'm kind of struggling with the expression to fill in. Each post's `authors` attribute is a list of objects that have first names and last names. Is there a way I can check if an object with a specific first name and last name is in the list?

**Comment by malsf21** [comment]
Yeah, this is a good point - in hindsight, I do wish we designed team pages around a single name (e.g. `Matthew Wang`) instead of a first name / last name. Maybe I'll ping @krashanoff to see if he has any thoughts on this? Worst-case, I'll give this a once over and then just let it merge - may mean that we're due for another refactor on how we handle names.

**Comment by MDZHX** [comment]
Thanks! That would be helpful :)

**Comment by krashanoff** [comment]
> in hindsight, I do wish we designed team pages around a single name (e.g. Matthew Wang) instead of a first name / last name.

This is what I was originally trying for when I first implemented team pages, but I remember that after some discussion we switched to separating out first and last names. I think it was for sorting members on the team page by name. There may also have been an issue with naming ambiguities.

To help you along with your expression trouble, what if you checked that each first name and last name field matched? For example, assuming that you have access to the author information in your post, or pass it down through some headmatter (accessible via the `page` variable:

```
{% assign authors_posts = site.posts | where_exp:"post", "post.fname == page.fname and post.lname == page.lname" | sort_by:"date" %}
```

> may mean that we're due for another refactor on how we handle names.

We could refactor posts specifically to just reference authors by their team file (e.g., `mwang` corresponds to `_team/mwang.md`). Thoughts?

**Comment by malsf21** [comment]
> We could refactor posts specifically to just reference authors by their team file (e.g., mwang corresponds to _team/mwang.md). Thoughts?

This could be something we tackle later, can you index collections by filename? I've actually ... never tried, haha.

For now, let's limit the scope of this PR. If you're able to get this to work Zihan, that would be appreciated, but if not I'll give your live demo one more test and give it the go-ahead to merge!

**Comment by MDZHX** [comment]
I tried Leo's quick fix but that doesn't seem to work based on the current `post.authors` structure, since `authors` is a collection of objects that also needs to be iterated over in order to find an object with the same `.fname` and `.lname` attributes. I'll do some more research to see if there's a fix, and if not I guess we can merge this one first and change the implementation after a possible refactoring of posts.

**Comment by MDZHX** [comment]
It seems that liquid doesn't have an elegant built-in way to check for an object in an array of objects.

I also tried to nest the filters like this:
`{% assign authors_posts = site.posts | where_exp:"post", "{{ post.authors | find_exp:"author","author.fname == page.firstname and author.lname == page.lastname" }} == nil" | sort_by:"date" %}`
and found that liquid doesn't allow that.

So for now, I'm thinking of two ways to implement this without refactoring posts:
* One way is to first manually create an iterable `authors_posts` using nested for loops + if conditions, and then check its size/iterate through it to render the cards. 
* Alternatively, we can use the current boolean flag version.

I wonder what are your thoughts on this? @malsf21 Thanks!

**Review by malsf21 — APPROVED** [review]
Just talked to Zihan, this looks good to merge! We'll look for refactors + fixing more info in another PR.

**Pull request merged**